### PR TITLE
Move symbolic functions out of `ntheory`

### PR DESCRIPTION
**Pull request by haru-44**
#### References to other Issues or PRs

* #26034
* #26042

#### Brief description of what is fixed or changed

The following functions have been changed (`mobius` has been fixed in other PRs).

|functions                                                                         |states     |target location                                        |new/existing |imports                 |
|:-------------------------------------------------------------------------------- |:--------- |:----------------------------------------------------- |:----------- |:---------------------- |
|sympy.ntheory.factor_.divisor_sigma                                               |deprecated |sympy.functions.combinatorial.numbers.divisor_sigma    |new          |top-level               |
|sympy.ntheory.factor_.primenu                                                     |deprecated |sympy.functions.combinatorial.numbers.primenu          |new          |top-level               |
|sympy.ntheory.factor_.primeomega                                                  |deprecated |sympy.functions.combinatorial.numbers.primeomega       |new          |top-level               |
|sympy.ntheory.factor_.reduce_totient                                              |deprecated |sympy.functions.combinatorial.numbers.reduce_totient   |new          |top-level               |
|sympy.ntheory.factor_.totient                                                     |deprecated |sympy.functions.combinatorial.numbers.totient          |new          |top-level               |
|sympy.ntheory.factor_.udivisor_sigma                                              |deprecated |sympy.functions.combinatorial.numbers.udivisor_sigma   |new          |`sympy.functions`       |
|sympy.ntheory.generate.primepi                                                    |deprecated |sympy.functions.combinatorial.numbers.primepi          |new          |top-level               |
|sympy.partitions_.npartitions                                                     |deprecated |sympy.functions.combinatorial.numbers.partition        |existing     |top-level               |
|sympy.ntheory.residue_ntheory.jacobi_symbol                                       |deprecated |sympy.functions.combinatorial.numbers.jacobi_symbol    |new          |top-level               |
|sympy.ntheory.residue_ntheory.kronecker_symbol                                    |deleted    |sympy.functions.combinatorial.numbers.kronecker_symbol |new          |top-level               |
|sympy.ntheory.residue_ntheory.legendre_symbol                                     |deprecated |sympy.functions.combinatorial.numbers.legendre_symbol  |new          |top-level               |
|sympy.ntheory.residue_ntheory.mobius                                              |deprecated |sympy.functions.combinatorial.numbers.mobius           |new          |top-level               |
|sympy.functions.combinatorial.numbers.carmichael.is_carmichael                    |deprecated |sympy.ntheory.factor_.is_carmichael                    |new          |top-level               |
|sympy.functions.combinatorial.numbers.carmichael.find_carmichael_numbers_in_range |deprecated |sympy.ntheory.factor_.find_carmichael_numbers_in_range |new          |`sympy.ntheory.factor_` |
|sympy.functions.combinatorial.numbers.carmichael.find_first_n_carmichaels         |deprecated |sympy.ntheory.factor_.find_first_n_carmichaels         |new          |`sympy.ntheory.factor_` |

* Most functions are `deprecated`, but the `kronecker_symbol` function was added to `ntheory` after the 1.12 release and deleted.
* Most functions are newly created. Since `npartitions` and `partition` existed as functions with similar functions, they have been consolidated into `partition`.
* Functions marked `top-level` can be imported like `from sympy import mobius`. Other functions can be imported as described (e.g., `from sympy.functions import udivisor_sigma`).
* The following functions have been added as private functions. All of them are functions for numerical calculations.
  * `sympy.ntheory.factor_._divisor_sigma`
  * `sympy.ntheory.generate._primepi`
  * `sympy.partitions_._partition_rec`, `sympy.partitions_._partition`

#### Release Notes

* functions
  * DEPRECATION : moved `is_carmichael` from `functions` to `ntheory`. The code `from sympy import carmichael; carmichael.is_carmichael(1)` or `from sympy.functions import carmichael; carmichael.is_carmichael(1)` needs to be modified to `from sympy import is_carmichael; is_carmichael(1)`.
  * DEPRECATION : moved `find_carmichael_numbers_in_range`, `find_first_n_carmichaels` from `functions` to `ntheory`. The code `from sympy import carmichael; carmichael.find_carmichael_numbers_in_range(1, 2)` or `from sympy.functions import carmichael; carmichael.find_carmichael_numbers_in_range(1, 2)` needs to be modified to `from sympy.ntheory.factor_ import find_carmichael_numbers_in_range; find_carmichael_numbers_in_range(1, 2)`.
  * Add new function `kronecker_symbol`
* ntheory
  * DEPRECATION : moved `primenu`, `primeomega`, `totient`, `reduce_totient`, `divisor_sigma`, `primepi`, `npartitions`, `legendre_symbol`, `jacobi_symbol` from `ntheory` to `functions`. Code that imports these functions from top-level like ``from sympy import mobius`` will continue to work fine. However code that imports these from the fully qualified module like ``from sympy.ntheory import mobius`` or ``from sympy.ntheory.residue_ntheory import mobius`` will now see a deprecation warning. The new location for these functions is in ``sympy.functions`` but the intended way to import them is still from top-level like ``from sympy import mobius``.
  * DEPRECATION : moved `udivisor_sigma` from `ntheory` to `functions`. The code `from sympy.ntheory.factor_ import udivisor_sigma` needs to be modified to `from sympy.functions import udivisor_sigma`.

### CODE REVIEW

File: doc/src/explanation/active-deprecations.md [hunk 1]
```diff
@@ -435,6 +435,9 @@ than an element method.
 Symbolic functions in ``ntheory`` have been moved to ``functions``.
 Specifically, the following functions.
 
+* ``sympy.ntheory.factor_.primenu``
+* ``sympy.ntheory.factor_.primeomega``
+* ``sympy.ntheory.factor_.totient``
 * ``sympy.ntheory.residue_ntheory.mobius``
```
Comment (oscarbenjamin) [inline doc/src/explanation/active-deprecations.md:441]: The deprecation note here should be a bit clearer about what is happening from a user perspective.

All of these functions can be imported like `from sympy import mobius` which continues to work just fine after these changes as well as it did before. What is relevant for users/downstream is that code that does `from sympy.ntheory.residue_ntheory import mobius` should be changed to `from sympy import mobius`.

Comment (haru-44) [reply doc/src/explanation/active-deprecations.md]: I added a description.

File: doc/src/explanation/active-deprecations.md [hunk 2]
```diff
@@ -435,6 +435,14 @@ than an element method.
...
 
+All of these functions can be imported like ``from sympy import mobius`` which
+continues to work just fine after these changes as well as it did before.
+However, if you are using ``from sympy.ntheory import mobius``, you need to
+change to ``from sympy import mobius``.
+
+* ``sympy.ntheory.factor_.primenu``
+* ``sympy.ntheory.factor_.primeomega``
+* ``sympy.ntheory.factor_.totient``
 * ``sympy.ntheory.residue_ntheory.mobius``
```
Comment (oscarbenjamin) [inline doc/src/explanation/active-deprecations.md:446]: The list of functions should come after "the following functions".

Comment (haru-44) [reply doc/src/explanation/active-deprecations.md:446]: Is this what you mean?

```
Symbolic functions in ``ntheory`` have been moved to ``functions``.
All of these functions can be imported like ``from sympy import mobius`` which
continues to work just fine after these changes as well as it did before.
However, if you are using ``from sympy.ntheory import mobius``, you need to
change to ``from sympy import mobius``. Specifically, the following functions.

* ...
```

Comment (oscarbenjamin) [reply doc/src/explanation/active-deprecations.md:446]: Yes, that is one possibility. Just when you say "the following functions" usually the list should follow immediately.

File: doc/src/explanation/active-deprecations.md [hunk 3]
```diff
@@ -433,8 +433,14 @@ than an element method.
...
-
+All of these functions can be imported like ``from sympy import mobius`` which
+continues to work just fine after these changes as well as it did before.
+However, if you are using ``from sympy.ntheory import mobius``, you need to
+change to ``from sympy import mobius``. Specifically, the following functions.
+
+* ``sympy.ntheory.factor_.primenu``
+* ``sympy.ntheory.factor_.primeomega``
+* ``sympy.ntheory.factor_.totient``
 * ``sympy.ntheory.residue_ntheory.mobius``
```
Comment (oscarbenjamin) [inline doc/src/explanation/active-deprecations.md:444]: Okay here is how I suggest to word the whole thing:
```
### Relocate symbolic functions from ``ntheory`` to ``functions``

The following symbolic functions in ``ntheory`` have been moved to
``functions``:

* ``sympy.ntheory.factor_.primenu``
* ``sympy.ntheory.factor_.primeomega``
* ``sympy.ntheory.factor_.totient``
* ``sympy.ntheory.residue_ntheory.mobius``

Code that imports these functions from top-level like ``from sympy import
mobius`` will continue to work fine. However code that imports these from the
fully qualified module like ``from sympy.ntheory import mobius`` or ``from
sympy.ntheory.residue_ntheory import mobius`` will now see a deprecation
warning. The new location for these functions is in ``sympy.functions`` but the
intended way to import them is still from top-level like ``from sympy import
mobius``.
```

Comment (haru-44) [reply doc/src/explanation/active-deprecations.md]: Thank you.

**Comment by oscarbenjamin** [comment]
Can you include a summary in the OP of what happened with each individual function like what has been moved where, what has been renamed or deprecated, what has been added etc?

Are all of the functions that have been moved importable at top-level in SymPy 1.12?

The release note should say DEPRECATION rather than BREAKING CHANGE if something is deprecated but still works. The release note should also be clear about how downstream code should be expected to be changed as a result of the changes here.

**Comment by oscarbenjamin** [comment]
I would like someone else to review this.

**Comment by haru-44** [comment]
> Can you include a summary in the OP of what happened with each individual function like what has been moved where, what has been renamed or deprecated, what has been added etc?
> 
> Are all of the functions that have been moved importable at top-level in SymPy 1.12?
> 
> The release note should say DEPRECATION rather than BREAKING CHANGE if something is deprecated but still works. The release note should also be clear about how downstream code should be expected to be changed as a result of the changes here.

Fixed the OP. If there are any unclear points, please let me know.

**Comment by haru-44** [comment]
If possible, I would like to get this done in time for the 1.13 release(#26169).

**Comment by oscarbenjamin** [comment]
Looks good. Thanks @haru-44 !

**Comment by haru-44** [comment]
thanks

**Pull request merged**